internal/faults/common.go

```diff
@@ -0,0 +1,278 @@
+// Package faults provides common headers and client-side fault injection
+// functionality.
+package faults
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"math/rand/v2"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/reddit/baseplate.go/internal/prometheusbpint"
+)
+
+const (
+	promNamespace      = "faultbp"
+	clientNameLabel    = "fault_client_name"
+	serviceLabel       = "fault_service"
+	methodLabel        = "fault_method"
+	protocolLabel      = "fault_protocol"
+	successLabel       = "fault_success"
+	delayInjectedLabel = "fault_injected_delay"
+	abortInjectedLabel = "fault_injected_abort"
+)
+
+var (
+	TotalRequests = promauto.With(prometheusbpint.GlobalRegistry).NewCounterVec(prometheus.CounterOpts{
+		Namespace: promNamespace,
+		Name:      "fault_requests_total",
+		Help:      "Total count of requests seen by the fault injection middleware.",
+	}, []string{
+		clientNameLabel,
+		serviceLabel,
+		methodLabel,
+		protocolLabel,
+		successLabel,
+		delayInjectedLabel,
+		abortInjectedLabel,
+	})
+)
+
+// Headers is an interface to be implemented by the caller to allow
+// protocol-specific header lookup. Using an interface here rather than a
+// function type avoids any potential closure requirements of a function.
+type Headers interface {
+	// Lookup returns the value of a protocol-specific header with the
+	// given key.
+	Lookup(ctx context.Context, key string) (string, error)
+}
+
+// ResumeFn is the function type to continue processing the protocol-specific
+// request without injecting a fault.
+type ResumeFn[T any] func() (T, error)
+
+// FaultFn is the function type to inject a protocol-specific fault with the
+// given code and message.
+type FaultFn[T any] func(code int, message string) (T, error)
```

Comment:
Similarly here, maybe `type Resume` and `type Fault` is as clear. At least Resume already describe an action. Where as Fault should maybe be named something with `Inject`?

Reply:
Updated to `Resume` and `Abort`.